### ISSUE TITLE
Add official support for thread IDs

### ISSUE THREAD

**Issue by gdonval** [numba/numba#7936]
---

## Feature request

A lot of the current pain points in parallel mode could be alleviated if one had access to some sort of unique thread ID within a specific parallel context.

As per the documentation, `numba.np.ufunc.parallel._get_thread_id()` does this **but** it is private and the documentation warns against its use in any actual code besides tests.

**Comment by sklam** [comment]
@stuartarchibald @DrTodd13, do you remember the details why we don't expose thread IDs yet?

**Comment by DrTodd13** [comment]
My recollection is that I had to do something in the autochunk PR to guarantee not only that we have thread IDs but that they are in the range 0-N.  I'm of the opinion that we should have this guarantee and expose it publicly.

**Comment by stuartarchibald** [comment]
Duplicate of https://github.com/numba/numba/issues/3130?

I think there's multiple attempts to expose the thread ID, @DrTodd13 added one in https://github.com/numba/numba/pull/7625 and I recollect adding one as part of https://github.com/numba/numba/pull/6674 (built on top of @DrTodd13's PR https://github.com/numba/numba/pull/6025) in patch https://github.com/numba/numba/pull/6674/commits/35d2b1051456255e9d10fb9de8f0027012690f72. I also seem to remember discussing it at a Numba public meeting and the conclusion being that it can to made public so long as the behaviour is consistent across threadpool implementations (i.e. IDs are in range 0-N, but with no guarantee of all ID being present).
